### PR TITLE
refactor: use temp table to refactor materialized cte

### PR DESCRIPTION
**Pull request by xudong963**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

Use temp table to refactor materialized cte:

1. We can eliminate many unnecessary codes and improve maintenance.
2. Add the setting `persist_materialized_cte` to decide if the cte is to persist storage or not. The default behavior is using the memory table. If the user's materialized cte results are large and may OOM, they can open the setting.
3. Support distribution materialized cte. TPCH 15 under cluser may be faster, but ci-benchmark is broken, will check later

- fixes: https://github.com/databendlabs/databend/issues/16874 https://github.com/databendlabs/databend/issues/16858 https://github.com/databendlabs/databend/issues/16832

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [x] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/databendlabs/databend/16900)

### CODE REVIEW

**Comment by xudong963** [comment]
Wait https://github.com/databendlabs/databend/issues/16941

**Comment by xudong963** [comment]
```
12_0005_changes_select:                                                 [ FAIL ] - result differs with:
--- /runner/_work/databend/databend/tests/suites/0_stateless/12_time_travel/12_0005_changes_select.result	2024-11-25 12:23:54.553130727 +0000
+++ /runner/_work/databend/databend/tests/suites/0_stateless/12_time_travel/12_0005_changes_select.stdout	2024-11-25 12:26:02.435793713 +0000
@@ -1,22 +1,15 @@
 latest snapshot should contain 2 rows
 2
 changes default snapshot at the first insertion
-1	1	DELETE	false
-2	1	DELETE	true
-2	2	INSERT	true
-3	3	INSERT	false
+Error: APIError: ResponseError with 1006: can not use temp table in http handler if cookie is not enabled
 changes default snapshot at the first insertion end the deletion
-1	1	DELETE	false
-2	1	DELETE	true
-2	2	INSERT	true
+Error: APIError: ResponseError with 1006: can not use temp table in http handler if cookie is not enabled
 changes append_only snapshot at the first insertion
 ```
 **can not use temp table in http handler if cookie is not enabled**

http session for temp table has supported, is there a plan to update client? cc @everpcpc @sundy-li

**Comment by sundy-li** [comment]
> http session for temp table has supported, is there a plan to update client? 

@youngsofun  to support it.

**Comment by everpcpc** [comment]
bendsql in runner upgraded to 0.23.1

**Comment by xudong963** [comment]
It seems that ci benchmark is broken. cc @everpcpc

**Comment by everpcpc** [comment]
> It seems that ci benchmark is broken. cc @everpcpc

cc @zhang2014

File: src/query/service/src/sessions/query_ctx_shared.rs [hunk 1]
```diff
@@ -361,6 +357,16 @@ impl QueryContextShared {
         Ok(res)
     }
 
+    pub fn remove_table_from_cache(&self, catalog_name: &str, database: &str, table: &str) {
```
Comment (SkyFan2002) [inline src/query/service/src/sessions/query_ctx_shared.rs:360]: This is an existing function `evict_table_from_cache()` similar to this.

File: src/query/service/src/interpreters/interpreter.rs [hunk 2]
```diff
@@ -85,7 +85,10 @@ pub trait Interpreter: Sync + Send {
     #[fastrace::trace]
     async fn execute(&self, ctx: Arc<QueryContext>) -> Result<SendableDataBlockStream> {
         log_query_start(&ctx);
-        match self.execute_inner(ctx.clone()).await {
+        let res = self.execute_inner(ctx.clone()).await;
```
Comment (SkyFan2002) [inline src/query/service/src/interpreters/interpreter.rs:88]: I'm not sure if the pipeline finishes executing here. It looks like `execute_inner()` returns a `SendableDataBlockStream` and needs to be polled later. Maybe we should call `ctx.drop_m_cte_temp_table()` with `pipeline.set_on_finished()`.

File: src/query/sql/src/planner/binder/bind_query/bind.rs [hunk 3]
```diff
@@ -165,4 +178,54 @@ impl Binder {
...
+            .ctx
+            .is_temp_table(CATALOG_DEFAULT, &database, &table_name)
+        {
+            return Err(ErrorCode::Internal(format!(
+                "Temporary table {:?} already exists in current session, please change the materialized CTE name",
+                table_name
+            )));
+        }
+        let create_table_stmt = CreateTableStmt {
+            create_option: CreateOption::CreateOrReplace,
```
Comment (SkyFan2002) [inline src/query/sql/src/planner/binder/bind_query/bind.rs:201]: Just `CreateOption::Create` is ok.

File: tests/sqllogictests/suites/mode/standalone/explain/materialized_cte.test [hunk 4]
```diff
@@ -1,61 +1,57 @@
...
-    ├── probe keys: [t1.a (#1)]
-    ├── filters: []
-    ├── estimated rows: 100.00
-    ├── CTEScan(Build)
-    │   ├── CTE index: 0, sub index: 2
-    │   └── estimated rows: 10.00
-    └── CTEScan(Probe)
-        ├── CTE index: 0, sub index: 1
-        └── estimated rows: 10.00
+│   └── estimated rows: 0.00
```
[the remark this answers is not in the document]
Comment (xudong963) [reply tests/sqllogictests/suites/mode/standalone/explain/materialized_cte.test:19]: Yes, seems that memory table doesn't have statistics, we need to add for it.

[the remark this answers is not in the document]
Comment (xudong963) [reply tests/sqllogictests/suites/mode/standalone/explain/materialized_cte.test]: add table stats for memory table: https://github.com/databendlabs/databend/pull/16900/commits/5127147a834195038c3a2e240185dcf38eb2de33, let's try to add col stats later: https://github.com/databendlabs/databend/issues/16994

**Review by Dousir9 — APPROVED** [review]

**Review by SkyFan2002 — APPROVED** [review]

**Pull request merged**